src/devices/src/virtio/mod.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+// We're only providing virtio over MMIO devices for now, but we aim to add PCI support as well.
+
+pub mod block;
+
+use std::sync::atomic::{AtomicU8, Ordering};
+use std::sync::{Arc, Mutex};
+
+use event_manager::{EventManager, MutEventSubscriber};
+use kvm_ioctls::VmFd;
+use vm_device::bus::MmioRange;
+use vmm_sys_util::eventfd::EventFd;
+
+// TODO: Move virtio-related defines from the local modules to the `vm-virtio` crate upstream.
+
+// TODO: Add MMIO-specific module when we add support for something like PCI as well.
+
+// Device-independent virtio features.
+mod features {
+    pub const VIRTIO_F_RING_EVENT_IDX: u64 = 29;
+    pub const VIRTIO_F_VERSION_1: u64 = 32;
+    pub const VIRTIO_F_IN_ORDER: u64 = 35;
+}
+
+// This bit is set on the device interrupt status when notifying the driver about used
+// queue events.
+// TODO: There seem to be similar semantics when the PCI transport is used with MSI-X cap
+// disabled. Let's figure out at some point if having MMIO as part of the name is necessary.
+const VIRTIO_MMIO_INT_VRING: u8 = 0x01;
+
+// The driver will write to the register at this offset in the MMIO region to notify the device
+// about available queue events.
+const VIRTIO_MMIO_QUEUE_NOTIFY_OFFSET: u64 = 0x50;
+
+// TODO: Make configurable for each device maybe?
+const QUEUE_MAX_SIZE: u16 = 256;
+
+#[derive(Copy, Clone)]
+pub struct MmioConfig {
+    pub range: MmioRange,
+    // The interrupt assigned to the device.
+    pub gsi: u32,
+}
+
+// These arguments are common for all virtio devices. We're always passing a mmio_cfg object
+// for now, and we'll re-evaluate the layout of this struct when adding more transport options.
+pub struct CommonArgs<'a, M, B> {
```

Comment:
I wonder if we can find another word instead of `args` (applies to `BlockArgs` as well). `args` makes me think of cmd line arguments :-?

Reply:
I tried to think about this as well, but didn't come up with a nicer option ('Params` or something like that is another, but not great either). Ultimately we can rename this as soon as we can think of something else.